File: onboard/DroneBrain/app/src/main/java/edu/cmu/cs/dronebrain/interfaces/DroneItf.java
package edu.cmu.cs.dronebrain.interfaces;

import java.util.Vector;

public interface DroneItf {
    void init() throws Exception;
    void connect() throws Exception;
    void disconnect() throws Exception;
    void takeOff() throws Exception;
    void land() throws Exception;
    void setHome(Double lat, Double lng) throws Exception;
    void moveTo(Double lat, Double lng, Double alt) throws Exception;
    void moveBy(Double x, Double y, Double z) throws Exception; //x,y,z in meters
    void startStreaming(Integer resolution) throws Exception;
    byte[] getVideoFrame() throws Exception;
    void stopStreaming() throws Exception;
    void rotateBy(Double theta) throws Exception;
    void rotateTo(Double theta) throws Exception;
    void setGimbalPose(Double yaw_theta, Double pitch_theta, Double roll_theta) throws Exception;
    void takePhoto() throws Exception;
    void trackTarget(Vector<Double> box, Double leash) throws Exception;
    void getStatus() throws Exception;
    String getName() throws Exception;
    Double getLat() throws Exception;
    Double getLon() throws Exception;
    Double getAlt() throws Exception;
    void hover() throws Exception;
    void cancel() throws Exception;
    void kill() throws Exception;
}

File: hermes/python/interfaces/DroneItf.py
# SPDX-FileCopyrightText: 2023 Carnegie Mellon University - Satyalab
#
# SPDX-License-Identifier: GPL-2.0-only
import asyncio
from abc import ABC, abstractmethod

class DroneItf(ABC):

    ''' Connection methods '''
    
    @abstractmethod
    async def connect(self):
        pass

    @abstractmethod
    async def isConnected(self):
        pass

    @abstractmethod
    async def disconnect(self):
        pass
    
    ''' Streaming methods '''
    
    @abstractmethod
    async def startStreaming(self, **kwargs):
        pass
    
    @abstractmethod
    async def getVideoFrame(self):
        pass
    
    @abstractmethod
    async def stopStreaming(self):
        pass

    ''' Take off/ Landing methods '''

    @abstractmethod
    async def takeOff(self):
        pass

    @abstractmethod
    async def land(self):
        pass

    @abstractmethod
    async def setHome(self, lat, lng):
        pass

    @abstractmethod
    async def rth(self):
        pass
    
    ''' Movement methods '''

    @abstractmethod
    async def PCMD(self, pitch, yaw, roll, gaz, rot):
        pass

    @abstractmethod
    async def moveTo(self, lat, lng, alt):
        pass
    
    @abstractmethod
    async def moveBy(self, x, y, z, t):
        pass

    @abstractmethod
    async def rotateTo(self, theta):
        pass

    @abstractmethod
    async def setGimbalPose(self, yaw_theta, pitch_theta, roll_theta):
        pass

    @abstractmethod
    async def hover(self):
        pass

    ''' Photography methods ''' 

    @abstractmethod
    async def takePhoto(self):
        pass

    @abstractmethod
    async def toggleThermal(self, on):
        pass

    ''' Status methods '''

    @abstractmethod
    async def getName(self):
        pass

    @abstractmethod
    async def getLat(self):
        pass

    @abstractmethod
    async def getLng(self):
        pass
    
    @abstractmethod
    async def getHeading(self):
        pass

    @abstractmethod
    async def getRelAlt(self):
        pass

    @abstractmethod
    async def getExactAlt(self):
        pass
    
    @abstractmethod
    async def getRSSI(self):
        pass

    @abstractmethod
    async def getBatteryPercentage(self):
        pass

    @abstractmethod
    async def getMagnetometerReading(self):
        pass

    @abstractmethod
    async def getSatellites(self):
        pass

    @abstractmethod
    def getGimbalPitch(self):
        pass
    
    ''' Control methods '''

    @abstractmethod
    async def kill(self):
        pass
